c10/core/TensorImpl.h

```diff
@@ -138,6 +138,68 @@ struct C10_API AutogradMetaInterface {
   virtual ~AutogradMetaInterface();
 };
 
+// NOTE [ Version Counter Sharing ]
+//
+// Every Tensor has a version counter. Version counters are incremented
+// whenever the data or shape of a tensor changes through Variable operations.
+// These are typically in-place operations. Version counters are used to
```

Comment:
I don't think it's just typical: the *only* way we should be bumping the version counter, is if we do an in-place mutation.